src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+            oob.set_big_data_receiver_handle(raw_receiver_handle);
+        }
+
+        // If we need to send OOB data, serialize it
+        let mut oob_data: Vec<u8> = vec![];
+        if oob.needs_to_be_sent() {
+            oob_data = bincode::serde::serialize(&oob, bincode::SizeLimit::Infinite).unwrap();
+        }
+
+        unsafe {
+            // if we need to use the fragment channel, then we send a 0 as the data size;
+            // the receiver will see that there's a big data receiver in the OOB portion
+            let header: [u32; 2] = [
+                if big_data_sender.is_some() { 0 } else { data.len() } as u32,
+                oob_data.len() as u32
+            ];
+            let header_bytes: &[u8] = slice::from_raw_parts(header.as_ptr() as *const u8, HEADER_SIZE);
+
+            // We need to write the main message first, which will indicate to the receiver
+            // that it should receive a message on the OOB big data channel.  If we don't
+            // send this first we'll block
+            try!(write_buf(*self.handle, &header_bytes));
+            if big_data_sender.is_none() {
+                try!(write_buf(*self.handle, data));
+            }
+            try!(write_buf(*self.handle, &oob_data));
+            if big_data_sender.is_some() {
+                dd2!("[c {:?}] sending unfragmented to big-data channel: {} bytes",
+                     *self.handle, data.len());
+                try!(big_data_sender.unwrap().send_unfragmented(data));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub struct OsIpcReceiverSet {
+    // the IOCP that we select on
+    iocp: WinHandle,
+
+    // the set of receivers
+    receivers: Vec<OsIpcReceiver>,
+}
+
+pub enum OsIpcSelectionResult {
+    DataReceived(i64, Vec<u8>, Vec<OsOpaqueIpcChannel>, Vec<OsIpcSharedMemory>),
+    ChannelClosed(i64),
+}
+
+impl OsIpcReceiverSet {
+    pub fn new() -> Result<OsIpcReceiverSet,WinError> {
+        unsafe {
+            let iocp = kernel32::CreateIoCompletionPort(INVALID_HANDLE_VALUE,
+                                                        ptr::null_mut(),
+                                                        0 as winapi::ULONG_PTR,
+                                                        0);
+            if iocp.is_null() {
+                return Err(WinError::last("CreateIoCompletionPort"));
+            }
+
+            Ok(OsIpcReceiverSet {
+                iocp: WinHandle::new(iocp),
+                receivers: vec![],
+            })
+        }
+    }
+
+    pub fn add(&mut self, mut receiver: OsIpcReceiver) -> Result<i64,WinError> {
+        // use this to identify the receiver
+        let receiver_handle = *receiver.handle;
+
+        try!(receiver.add_to_iocp(*self.iocp));
+
+        // FIXME we *may* need a mutex to protect self.receivers --
+        // one thread could be adding something to this Set while
+        // another is calling select(); the add() could cause the
```

Comment:
`Sync` was my first though as well (as you can see in my earlier, bogus comment on that topic) -- but now I realised that this doesn't even come into play here, as `Sync` is only about *shared* references, in the face of interior mutability; while here we are actually using mutable references, without resorting to interior mutability at all...

(Note that in the *current* code, `OsIpcReceiverSet` is indeed `Sync`, since all the constituent types are -- regardless of the `Vec<>` container. This will change though when `OsIpcReceiver` is no longer marked `Sync`. But again, this doesn't even come into play here...)